go/kbcmf/decrypt.go

```diff
@@ -24,96 +25,107 @@ type Keyring interface {
 	LookupBoxPublicKey(kid []byte) BoxPublicKey
 }
 
-type publicDecryptStream struct {
-	output     io.Writer
+type state int
+
+const (
+	stateHeader      state = iota
+	stateBody        state = iota
+	stateEndOfStream state = iota
+)
+
+type decryptStream struct {
 	ring       Keyring
 	fmps       *framedMsgpackStream
 	err        error
-	state      int
+	state      state
 	keys       *receiverKeysPlaintext
 	sessionKey SymmetricKey
+	buf        []byte
 }
 
-func (pds *publicDecryptStream) Write(b []byte) (n int, err error) {
-	if pds.err != nil {
-		return 0, pds.err
+func (ds *decryptStream) Read(b []byte) (n int, err error) {
+	for n == 0 && err == nil {
+		n, err = ds.read(b)
 	}
-	if n, pds.err = pds.fmps.Write(b); pds.err != nil {
-		return 0, pds.err
+	if err == io.EOF && ds.state != stateEndOfStream {
+		err = io.ErrUnexpectedEOF
 	}
-	if pds.err = pds.decode(); pds.err != nil {
-		return 0, pds.err
-	}
-	return n, nil
+	return n, err
 }
 
-func (pds *publicDecryptStream) Close() (err error) {
-	if pds.err != nil {
-		return pds.err
+func (ds *decryptStream) read(b []byte) (n int, err error) {
+
+	if len(ds.buf) > 0 {
+		n = copy(b, ds.buf)
+		ds.buf = ds.buf[n:]
+		return n, nil
 	}
-	if pds.err = pds.fmps.Close(); pds.err != nil {
-		return pds.err
+
+	if ds.state == stateHeader {
+		if err = ds.readHeader(); err != nil {
+			return 0, err
+		}
+		ds.state++
 	}
-	pds.err = pds.decode()
-	return pds.err
-}
 
-func (pds *publicDecryptStream) decode() error {
-	var err error
-	var seqno PacketSeqno
+	if ds.state == stateEndOfStream {
+		return 0, ds.readEndOfStream()
```

Comment:
Max also says: Let's put this part at the end. That way we can check it after the last read automatically.